app/code/Magento/Ui/view/base/web/js/grid/masonry.js

```diff
@@ -0,0 +1,280 @@
+             * @param int
+             */
+            containerWidth: window.innerWidth,
+
+            /**
+             * Margin between images
+             * @param int
+             */
+            imageMargin: 20,
+
+            /**
+             * Maximum image height value
+             * @param int
+             */
+            maxImageHeight: 240,
+
+            /**
+             * The value is minimum image width to height ratio when container width is less than the key
+             * @param {Object}
+             */
+            containerWidthToMinRatio: {
+                640: 3,
+                1280: 5,
+                1920: 8
+            },
+
+            /**
+             * Default minimal image width to height ratio.
+             * Applied when container width is greater than max width in the containerWidthToMinRatio matrix.
+             * @param int
+             */
+            defaultMinRatio: 10
+        },
+
+        /**
+         * Init observable variables
+         * @return {Object}
+         */
+        initObservable: function () {
+            this._super()
+                .observe([
+                    'rows',
+                    'errorMessage'
+                ]);
+
+            return this;
+        },
+
+        /**
+         * Init component handler
+         * @param {Object} rows
+         * @return {Object}
+         */
+        initComponent: function (rows) {
+            if (!rows || !rows.length) {
+                return;
+            }
+            this.imageMargin = parseInt(this.imageMargin, 10);
+            this.container = $('[data-id="' + this.containerId + '"]')[0];
+
+            this.setLayoutStyles();
+            this.setEventListener();
+
+            return this;
+        },
+
+        /**
+         * Set event listener to track resize event
+         */
+        setEventListener: function () {
+            var running = false,
+                handler = function () {
+                    this.containerWidth = window.innerWidth;
+                    this.setLayoutStyles();
+                }.bind(this);
+
+            window.addEventListener('resize', function () {
+                if (!running) {
+                    running = true;
+
+                    if (window.requestAnimationFrame) {
+                        window.requestAnimationFrame(function () {
+                            handler();
+                            running = false;
+                        });
+                    } else {
+                        setTimeout(function () {
+                            handler();
+                            running = false;
+                        }, 66);
+                    }
+                }
+            });
+        },
+
+        /**
+         * Set layout styles inside the container
+         */
+        setLayoutStyles: function () {
+            var containerWidth = parseInt(this.container.clientWidth, 10),
+                rowImages = [],
```

Comment:
The combination of a floating array (`rowImages`) and a `forEach` in the same area can be a good indication that `map` is better suited for the job. Would recommend switching over

Reply:
I don't think map can be used for this algorithm, rowImages is a temporary array containing images that should be in a single visual row on the screen. This array is flushed and recreated several times (= number of images rows) during the foreach execution. `this.rows().foreach` iterating over each data row (each image in this case)